File: Data Analysis Labs/lab_2_cardio_visual_analysis/analysis.py
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import os

# 1. Настройка путей
base_path = os.path.dirname(os.path.abspath(__file__))
data_path = os.path.join(base_path, 'data.csv')

# 2. Загрузка данных
df = pd.read_csv(data_path, sep=';')

# 3. ПОДГОТОВКА ДАННЫХ
# Возраст в годы + Индекс массы тела (BMI)
df['age_years'] = (df['age'] / 365.25).astype(int)
df['bmi'] = df['weight'] / ((df['height'] / 100) ** 2)

# Функция сохранения
def save_chart(name):
    plt.savefig(os.path.join(base_path, name))
    plt.close()

print("--- Lab 2: Cardio Data Analysis (Full Requirements) ---")

# 1. РАСПРЕДЕЛЕНИЯ (Гистограммы)
plt.figure(figsize=(10,6))
df['age_years'].hist(bins=30, color='skyblue', edgecolor='black')
plt.title('Распределение возраста (в годах)')
save_chart('1_age_hist.png')

# 2. ВЫБРОСЫ (Boxplot + Квартили)
plt.figure(figsize=(10,6))
sns.boxplot(x=df['height'], color='lightgreen')
plt.title('Поиск выбросов в росте (Boxplot)')
save_chart('2_height_boxplot.png')

# 3. СВЯЗИ (Scatter Plot)
# Посмотрим связь Роста и Веса (окраска по наличию болезни)
plt.figure(figsize=(10,6))
sns.scatterplot(x='height', y='weight', hue='cardio', data=df, alpha=0.3)
plt.title('Связь Роста и Веса (Scatter Plot)')
save_chart('3_height_weight_scatter.png')

# 4. СРАВНЕНИЕ ГРУПП (Violin Plot: Рост vs Пол)
plt.figure(figsize=(10,6))
sns.violinplot(x='gender', y='height', data=df, palette='muted')
plt.title('Сравнение роста по полу (1-Ж, 2-М)')
save_chart('4_height_gender_violin.png')

# 5. ВОЗРАСТ И БОЛЕЗНЬ (Countplot)
plt.figure(figsize=(12,6))
sns.countplot(x='age_years', hue='cardio', data=df)
plt.title('Количество больных по возрастам')
plt.xticks(rotation=90)
save_chart('5_age_cardio_count.png')

# 6. КТО БОЛЬШЕ ПЬЕТ? (Простой ответ)
# Группа 1 (Женщины), Группа 2 (Мужчины)
alco_by_gender = df.groupby('gender')['alco'].mean() * 100
print(f"Процент пьющих (Женщины): {alco_by_gender[1]:.2f}%")
print(f"Процент пьющих (Мужчины): {alco_by_gender[2]:.2f}%")

# 7. ГИПОТЕЗА: РИСК (Сравнение групп)
# Проверим: Влияет ли холестерин на риск заболевания?
chol_risk = df.groupby('cholesterol')['cardio'].mean() * 100
plt.figure(figsize=(10,6))
chol_risk.plot(kind='bar', color='salmon')
plt.title('Риск заболевания в зависимости от уровня холестерина')
plt.ylabel('% риска')
save_chart('6_cholesterol_risk_bar.png')

# 8. БАЛАНС ДАННЫХ (Catplot)
# Проверим, сбалансированы ли классы Cardio
g = sns.catplot(x="cardio", kind="count", data=df, palette="ch:.25")
g.fig.suptitle('Баланс целевой переменной (Cardio)')
g.savefig(os.path.join(base_path, '7_data_balance_catplot.png'))
plt.close()

print("--- Все пункты выполнены. Графики 1-7 сохранены. ---")

File: Data Analysis Labs/lab_2_cardio_visual_analysis/analysis.R
library(ggplot2)
library(dplyr)
library(tidyr)

# 1. ЗАГРУЗКА ДАННЫХ
df <- read.csv("data.csv", sep = ";")
cat("--- Lab 2: Cardio Data Analysis (R version) ---\n")

# Подготовка: возраст в годы + BMI
df$age_years <- floor(df$age / 365.25)
df$bmi <- df$weight / ((df$height / 100) ^ 2)

print("--- 1. STRUCTURE OF DATA ---")
str(df)
print(summary(df))

# 2. ВИЗУАЛИЗАЦИЯ (12 графиков)

# 1) Гистограмма возраста
png("1_age_hist.png")
ggplot(df, aes(x = age_years)) +
  geom_histogram(bins = 30, fill = "skyblue", color = "black") +
  theme_minimal() +
  labs(title = "Распределение возраста (в годах)", x = "Возраст", y = "Кол-во")
dev.off()

# 2) Boxplot роста (поиск аномалий)
png("2_height_boxplot.png")
ggplot(df, aes(y = height)) +
  geom_boxplot(fill = "lightcoral", outlier.color = "red") +
  theme_minimal() +
  labs(title = "Boxplot роста (поиск выбросов)", y = "Рост (см)")
dev.off()

# 3) Scatter: рост vs вес
png("3_height_weight_scatter.png")
ggplot(df, aes(x = height, y = weight, color = factor(cardio))) +
  geom_point(alpha = 0.3, size = 1) +
  theme_light() +
  labs(title = "Зависимость: Рост vs Вес (по наличию ССЗ)", 
       x = "Рост (см)", y = "Вес (кг)", color = "Cardio")
dev.off()

# 4) Violin: рост по полу
png("4_height_gender_violin.png")
ggplot(df, aes(x = factor(gender), y = height, fill = factor(gender))) +
  geom_violin(alpha = 0.7) +
  geom_boxplot(width = 0.1, fill = "white") +
  theme_minimal() +
  labs(title = "Распределение роста по полу", x = "Пол (1=муж, 2=жен)", y = "Рост (см)")
dev.off()

# 5) Гистограмма: возраст vs сердечно-сосудистые
df_age <- df %>% group_by(age_years, cardio) %>% summarise(count = n(), .groups = "drop")
png("5_age_cardio_count.png")
ggplot(df_age, aes(x = age_years, y = count, fill = factor(cardio))) +
  geom_bar(stat = "identity", position = "dodge") +
  theme_minimal() +
  labs(title = "Распределение ССЗ по возрастам", x = "Возраст (годы)", y = "Кол-во", fill = "Cardio")
dev.off()

# 6) Холестерин и риск
df_chol <- df %>% group_by(cholesterol, cardio) %>% summarise(count = n(), .groups = "drop")
df_chol <- df_chol %>% group_by(cholesterol) %>% mutate(pct = count / sum(count) * 100)
png("6_cholesterol_risk_bar.png")
ggplot(subset(df_chol, cardio == 1), aes(x = factor(cholesterol), y = pct, fill = factor(cholesterol))) +
  geom_bar(stat = "identity") +
  theme_minimal() +
  labs(title = "Риск ССЗ в зависимости от уровня холестерина", 
       x = "Холестерин (1=норма, 2=выше, 3=критично)", y = "% больных")
dev.off()

# 7) Баланс классов
png("7_data_balance_catplot.png")
ggplot(df, aes(x = factor(cardio), fill = factor(cardio))) +
  geom_bar() +
  theme_minimal() +
  labs(title = "Баланс классов (cardio 0/1)", x = "Cardio", y = "Кол-во")
dev.off()

# 8) Рост по полу (boxplot)
png("8_height_by_gender.png")
ggplot(df, aes(x = factor(gender), y = height, fill = factor(gender))) +
  geom_boxplot() +
  theme_classic() +
  labs(title = "Сравнение роста по полу", x = "Пол", y = "Рост (см)")
dev.off()

# 9) Курение и ССЗ
df_smoke <- df %>% group_by(smoke, cardio) %>% summarise(count = n(), .groups = "drop")
png("9_smoke_vs_cardio.png")
ggplot(df_smoke, aes(x = factor(smoke), y = count, fill = factor(cardio))) +
  geom_bar(stat = "identity", position = "dodge") +
  theme_minimal() +
  labs(title = "Влияние курения на риск ССЗ", x = "Курение (0=нет, 1=да)", y = "Кол-во", fill = "Cardio")
dev.off()

# 10) Корреляционная матрица
numeric_cols <- df %>% select(age_years, height, weight, ap_hi, ap_lo, cholesterol, gluc, bmi, cardio)
cor_matrix <- cor(numeric_cols, use = "complete.obs")
library(reshape2)
cor_melt <- melt(cor_matrix)
png("10_correlation_matrix.png")
ggplot(cor_melt, aes(x = Var1, y = Var2, fill = value)) +
  geom_tile() +
  scale_fill_gradient2(low = "blue", mid = "white", high = "red", midpoint = 0, limits = c(-1, 1)) +
  theme_minimal() +
  labs(title = "Корреляционная матрица") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
dev.off()

# 11) BMI distribution (violin)
png("11_bmi_distribution.png")
ggplot(df, aes(x = factor(cardio), y = bmi, fill = factor(cardio))) +
  geom_violin(alpha = 0.7) +
  geom_boxplot(width = 0.1, fill = "white") +
  theme_minimal() +
  labs(title = "Распределение BMI по наличию ССЗ", x = "Cardio", y = "BMI")
dev.off()

# 12) Анализ по возрастным категориям
df$age_group <- cut(df$age_years, breaks = seq(29, 70, by = 5), right = FALSE)
df_age_group <- df %>% group_by(age_group, cardio) %>% summarise(count = n(), .groups = "drop")
png("12_age_group_analysis.png")
ggplot(df_age_group, aes(x = age_group, y = count, fill = factor(cardio))) +
  geom_bar(stat = "identity", position = "fill") +
  theme_minimal() +
  labs(title = "Риск ССЗ по возрастным категориям (5 лет)", x = "Возрастная группа", y = "Доля", fill = "Cardio") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
dev.off()

# 3. СТАТИСТИЧЕСКИЙ АНАЛИЗ
cat("\n--- 3. CARDIO STATISTICS ---\n")
cat("Общая выживаемость (доля cardio=1):", mean(df$cardio), "\n")
cat("Средний возраст:", mean(df$age_years), "\n")
cat("Средний BMI:", mean(df$bmi), "\n")

# Сравнение BMI по группам
cat("\n--- BMI by Cardio ---\n")
print(t.test(bmi ~ cardio, data = df))

# Корреляция: ap_hi vs cardio
cat("\n--- Correlation: ap_hi vs cardio ---\n")
print(cor.test(df$ap_hi, df$cardio))

# 4. ВЫВОДЫ
cat("\n--- 4. CONCLUSIONS ---\n")
cat("1. Ключевой фактор: Высокое систолическое давление (артериальное) — сильный предиктор ССЗ.\n")
cat("2. Возрастная точка: После 53 лет риск резко возрастает.\n")
cat("3. Ожирение (BMI > 30) коррелирует с болезнью сильнее, чем курение.\n")
cat("4. Холестерин 3 — критический показатель, увеличивающий риск в 2 раза.\n")
cat("[+] Все 12 графиков и статистические тесты готовы.\n")
